internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -499,6 +511,44 @@ func (r *BMACReconciler) reconcileNodeLabels(bmh *bmh_v1alpha1.BareMetalHost, ag
 	return false
 }
 
+func (r *BMACReconciler) reconcileClusterReference(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost, agent *aiv1beta1.Agent, log logrus.FieldLogger) (bool, error) {
+	if clusterReferenceStr, ok := bmh.ObjectMeta.Annotations[BMH_CLUSTER_REFERENCE]; ok {
+		infraEnv, err := r.findInfraEnvForBMH(ctx, log, bmh)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to find infraenv for bmh %s/%s", bmh.Namespace, bmh.Name)
+		}
+		if infraEnv != nil && infraEnv.Spec.ClusterRef != nil {
+			return false, errors.Errorf("failed to reconcile cluster reference.  InfraEnv already has cluster reference %s/%s",
+				infraEnv.Spec.ClusterRef.Namespace, infraEnv.Spec.ClusterRef.Name)
+		}
+		var clusterReference aiv1beta1.ClusterReference
+		if err = json.Unmarshal([]byte(clusterReferenceStr), &clusterReference); err != nil {
+			return false, errors.Wrapf(err, "failed to reconcile cluster reference.  Unmarshal of cluster reference %s failed",
+				clusterReferenceStr)
+		}
+		if !reflect.DeepEqual(agent.Spec.ClusterDeploymentName, &clusterReference) {
+			agent.Spec.ClusterDeploymentName = &clusterReference
+			if agent.ObjectMeta.Annotations == nil {
+				agent.ObjectMeta.Annotations = make(map[string]string)
+			}
+			agent.ObjectMeta.Annotations[AGENT_CLUSTER_REFERENCE_MANAGED_BY_BMH] = "true"
```

Comment:
We have a helper for this [here](https://github.com/openshift/assisted-service/blob/2c149a3ddc43362414021dfc38598d3b9a8a467f/internal/controller/controllers/common.go#L385-L390)

Reply:
Also do we ever remove this annotation? I feel like managing two is harder than the suggestion I made in my previous comment.

Reply:
I don't see a reason to remove the annotation. It is an indication that the cluster-reference is managed by BMH, unless the user wants to use also the API against the agent - which doesn't make sense once it was already set by the BMH.

Assuming when using BMH as the CR to update, the cluster-reference is removed in the reverse of the method it was added.  Leaving the annotation empty seems like a strange way that is passed to the user.

Reply:
> We have a helper for this [here](https://github.com/openshift/assisted-service/blob/2c149a3ddc43362414021dfc38598d3b9a8a467f/internal/controller/controllers/common.go#L385-L390)

OK

Reply:
I just think you can accomplish the same thing with a single annotation rather than two. 

You only have three cases to deal with, right?

1. If a user wants the cluster ref managed by the BMH and want it bound they set the value of the annotation.
2. If they want it managed by the BMH and want it unbound they set the annotation to empty.
3. If they don't want the cluster ref managed by the BMH they remove the annotation.

This is also how the other annotations on the BMH work that currently control the agent spec, the only complication here should be that we're also ensuring there isn't already a cluster ref set on the infraenv.

Hostname, for example, doesn't have some annotation on the agent to indicate that it's managed by the BMH. And if you change the hostname annotation value to empty it unsets hostname, if you remove the annotation then hostname is no longer managed by BMH. I don't see why this annotation should work differently.

Reply:
Don't you think it is strange that a user added cluster-ref by adding an annotation that didn't exist before, and to remove it instead of returning it the way it was in the beginning, change it to a third form?  Seems strange to me.

Hostname is different.  There is reason to remove the hostname if it doesn't exist.  It doesn't make sense.  So it is one way update.   Foe cluster-ref it makes sense to remove the reference.

If you will look at node-labels you will be able to see that they are always synced.

Other than node-labels, there is no clearing operation on any other properties such as the cluster-reference.

Reply:
I think the way node labels works is wrong. If nothing is set on the BMH the user should be able to use the API on the agent the same as all the other annotation based options.

Reply:
For hostname, node labels, ignition config overrides, installation disk, etc., we override the value in the Agent as long as the value is set in the BMH.  If the user unsets the value in the BMH, we stop syncing it, but the value on the Agent remains.  There's no way to unset any of these properties on the Agent today except for unsetting it on both the BMH and the Agent.  This was OK until now because those properties are scoped to installation, so we didn't come across the case where a user set something on a BMH and then regretted it.  Even less so because we assumed people setting annotations on BMHs are using gitops and are more experienced.

So we have a gap here, but so far it hasn't bothered us.

The cluster ref property isn't scoped to installation - we assume that users will set and unset throughout the lifetime of the host.

Ideally we would unify the behavior for all annotations, but this changes the current behavior, so let's think about it past the scope of this PR.  I propose the behavior should be:
1. If we see MANAGED_BY_BMH= true, set or unset all of the Agent properties according to the BMH annotations.
2. If we see MANAGED_BY_BMH= false, ignore the BMH annotations.
3. If we see MANAGED_BY_BMH is not present on the Agent, and at least one of the annotations is set on the BMH, then set MANAGED_BY_BMH=true.

So if any of the annotations is set or unset on the BMH, all of the properties should be mirrored to the Agent (anything that was set or will be set on the Agent will be overwritten).  If the user wants to stop this behavior, they need to set MANAGED_BY_BMH=false.

As I said, for now, let's do it only for the cluster ref property.
@ori-amizur , this is close to what you implemented, but you're missing some of the behavior (e.g., #2 in the description above).

@mhrivnak Any thoughts?

Reply:
I'm okay with this approach as long as it's consistent and well communicated to the user.

If we only implement this for the cluster ref in this PR I'm worried no one will ever go and update the rest of the annotations to subscribe to the new method and we'd end up in this state forever which would be very unfortunate.

Reply:
My guess is that changing the behavior for all properties won't break anyone because users usually either make changes via BMH or Agent, and not both, but it's a risk.  WDYT?

Reply:
We'll have a better API and UX if the API user doesn't have to manage the behavior of controllers in addition to declaring desired state, and this `MANAGED_BY_BMH` value is really just changing the behavior of a controller. 

Do we break anyone if we just do this?
* if one of the annotations exists on the BMH, copy its value (even the empty value) to the agent.
* for annotations that don't exist on the BMH, don't do anything to corresponding values on the agent.

@ori-amizur I understand your concern that if I 1) add an annotation that has some effect, then 2) remove that annotation, it should un-do the effect. Unfortunately that's not how declarative APIs work. Whatever desired state is specified *right now*, the job of controllers is to ensure that the real state of the world matches that desired state. What happened or was desired in the past is irrelevant.

This discussion highlights the complexity of managing default values. If an API user unsets a field, should it be automatically set to a well-documented default? Or should the real-world state associated with that field be left alone? It depends on the specific field, but either way, fields that can be unset should have clear documentation about what will happen.

Then what should happen if someone deletes an annotation? Not only did they unset the field; they deleted the field itself! Obviously we can't set the field to a sensible default if the field no longer exists.

In this specific case, it's even more complex, since we're using annotations on API X as a pass-through to set desired state on API Y. If any defaulting is going to happen, it really should happen on API X, but we can't set defaults for annotations. I think it would be a mistake to add a setting on API Y that changes its behavior in an attempt to compensate for the uncertainties that come with using annotations on API X.

TL;DR let's please try to find a way to do this that doesn't care what happened in the past nor add an annotation that really just changes what assumptions a controller makes when it deals with unset values.